engine/collection/compliance/core_test.go

```diff
@@ -230,69 +253,152 @@ func (cs *CoreSuite) TestOnBlockProposalSkipProposalThreshold() {
+		cs.state.AssertNotCalled(cs.T(), "Extend", mock.Anything)
+		// we should not attempt to process the children
+		cs.pending.AssertNotCalled(cs.T(), "ByParentID", mock.Anything)
+	})
 }
 
-func (cs *CoreSuite) TestOnBlockProposalInvalidExtension() {
+// TestOnBlockProposal_InvalidExtension tests processing a proposal which passes HotStuff validation,
+// but fails compliance checks.
+//   - should not be added to the state
+//   - we should not attempt to process its children
+func (cs *CoreSuite) TestOnBlockProposal_InvalidExtension() {
 
 	// create a proposal that has two ancestors in the cache
 	originID := unittest.IdentifierFixture()
 	ancestor := unittest.ClusterBlockWithParent(cs.head)
 	parent := unittest.ClusterBlockWithParent(&ancestor)
 	block := unittest.ClusterBlockWithParent(&parent)
-	proposal := &messages.ClusterBlockProposal{
-		Header:  block.Header,
-		Payload: block.Payload,
-	}
+	proposal := unittest.ClusterProposalFromBlock(&block)
+	hotstuffProposal := model.ProposalFromFlow(block.Header, parent.Header.View)
 
 	// store the data for retrieval
 	cs.headerDB[parent.ID()] = &parent
 	cs.headerDB[ancestor.ID()] = &ancestor
 
-	// make sure we fail to extend the state
-	*cs.state = clusterstate.MutableState{}
-	cs.state.On("Final").Return(
-		func() clusterint.Snapshot {
-			return cs.snapshot
-		},
-	)
-	cs.state.On("Extend", mock.Anything).Return(errors.New("dummy error"))
+	// the block passes HotStuff validation
+	cs.validator.On("ValidateProposal", model.ProposalFromFlow(block.Header, parent.Header.View)).Return(nil)
 
-	// it should be processed without error
-	err := cs.core.OnBlockProposal(originID, proposal)
-	require.Error(cs.T(), err, "proposal with invalid extension should fail")
+	cs.Run("invalid block", func() {
+		// make sure we fail to extend the state
+		*cs.state = clusterstate.MutableState{}
+		cs.state.On("Final").Return(func() clusterint.Snapshot { return cs.snapshot })
+		cs.state.On("Extend", mock.Anything).Return(state.NewInvalidExtensionError(""))
+		// we should notify VoteAggregator about the invalid block
+		cs.voteAggregator.On("InvalidBlock", hotstuffProposal).Return(nil)
 
-	// we should extend the state with the header
-	cs.state.AssertCalled(cs.T(), "Extend", &block)
-
-	// we should not submit the proposal to hotstuff
-	cs.hotstuff.AssertExpectations(cs.T())
+		// the expected error should be handled within the Core
+		err := cs.core.OnBlockProposal(originID, proposal)
+		require.NoError(cs.T(), err, "proposal with invalid extension should fail")
+
+		// we should extend the state with the header
+		cs.state.AssertCalled(cs.T(), "Extend", &block)
+		// we should not pass the block to hotstuff
+		cs.hotstuff.AssertNotCalled(cs.T(), "SubmitProposal", mock.Anything, mock.Anything)
+		// we should not attempt to process the children
+		cs.pending.AssertNotCalled(cs.T(), "ByParentID", mock.Anything)
+		cs.voteAggregator.AssertExpectations(cs.T())
```

Comment:
a suggestion to use `mocks.NewVoteAggregator` and get rid of `AssertExpectations` in other places.

Reply:
👍 Was already using `NewVoteAggregator`, so `AssertExpectations` is indeed unnecessary - good point.